### PR TITLE
8306444: Don't leak memory in PhaseChaitin::PhaseChaitin

### PR DESCRIPTION
**Pull request by jdksjolen**
Hi,

First, `PhaseChaitin::PhaseChaitin` used to create 4 resource array of size `_cfg.number_of_blocks`: one to store all of the block pointers in (`_blks`), and three to do a sorting of the blocks in some order. The latter three weren't freed in the constructor, causing them to hang around for the entire duration of the phase. This is unnecessary, so this patch frees the arrays when we're done with them. It also allocates all of the resources arrays in one go.

Second, it copied over each partially filled bucket into the `_blks`, one block at a time. This patch changes this so that we don't allocate the `_blks` resource array at all, instead we simply squash all of the partially filled buckets into the first one using `::memmove`. 

I haven't done any micro benchmarking, but this should be faster and take less space.

This is currently passing tier1.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8306444](https://bugs.openjdk.org/browse/JDK-8306444): Don't leak memory in PhaseChaitin::PhaseChaitin

### Reviewers
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**)
 * [Roland Westrelin](https://openjdk.org/census#roland) (@rwestrel - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/13533/head:pull/13533` \
`$ git checkout pull/13533`

Update a local copy of the PR: \
`$ git checkout pull/13533` \
`$ git pull https://git.openjdk.org/jdk.git pull/13533/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 13533`

View PR using the GUI difftool: \
`$ git pr show -t 13533`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/13533.diff">https://git.openjdk.org/jdk/pull/13533.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/13533#issuecomment-1514723773)

### CODE REVIEW

File: src/hotspot/share/opto/chaitin.cpp
Comment (vnkozlov) [inline src/hotspot/share/opto/chaitin.cpp:265]: I did not know that you can free part of allocated space.

File: src/hotspot/share/opto/chaitin.cpp
Comment (vnkozlov) [inline src/hotspot/share/opto/chaitin.cpp:263]: May add assert that  `assert((offset - &buckets[0][0]) == nr_blocks`

File: src/hotspot/share/opto/chaitin.cpp
Comment (jdksjolen) [reply src/hotspot/share/opto/chaitin.cpp:265]: Arena and ResourceArea looks at array allocations as `sizeof(T)*N` bytes being allocated, nothing more than that. If it's the last thing you allocated then you can free that space.

File: src/hotspot/share/opto/chaitin.cpp
Comment (jdksjolen) [reply src/hotspot/share/opto/chaitin.cpp:263]: I added `assert((&buckets[0][0] + nr_blocks) == offset, "should be");`, it's easier for me to see that  `nr_blocks` is implicitly multiplied with `sizeof(Block*)`  :-).

**Review by vnkozlov — APPROVED** [review]
Looks good. You need second review for this.

**Review by rwestrel — APPROVED** [review]
Looks good to me.

**Pull request closed without merging**
